### DISCUSSION TITLE
Is it safe to call fmtlog::poll from the queue-full callback?

### DISCUSSION THREAD

**Discussion by gh-andre** [MengRao/fmtlog#63] [General]
Provided that `fmtlog::poll(true)` is called by a single thread from within the callback function set with `fmtlog::setLogQFullCB`, is it safe to call it there, like in the example below?

Asking in terms of reentrancy - would it interfere with any thread local storage, global values, etc, within `fmtlog`?

```c++
void logFullCB(void*)
{
    fmtlog::poll(true);
}

int main()
{
  fmtlog::setLogQFullCB(&logFullCB, nullptr);
  for (int i = 0; i < 1'000'000; i++) {
    std::string str(1000, ' ');
    logio("{:d} log q full cb test: {:s}", i, str);
  }
  return 0;
}
```

Thanks

**Comment by MengRao** [comment]
Yes, you may call `fmtlog::poll` in logFullCB, but be aware that `fmtlog::poll` is not thread-safe, so you should make sure it's not being called in another thread at the same time.
Also note that logFullCB is invoked synchronously in the `logio` macro in your example. In more complicated cases where multiple threads are using fmtlog for logging and logFullCB could be triggered concurrently, user should ensure thread-safety of `fmtlog::poll`.

**Reply by gh-andre** [reply]
This is just a simple example to demonstrate the intent. In actual code I do use synchronization to make sure `fmtlog::poll` is always called on a single thread. 

Initially I used a separate thread for this, but it takes a bit of time to synchronize between threads and calling `fmtlog::poll` right in the callback, with proper synchronization for multiple logging calls, is the most optimal way to flush the buffers. It works well, but I wanted to make sure I'm not missing anything. Thanks for confirming.
